File: typescript/src/generators/symbol.ts
/**
 * Unicode symbol generator.
 *
 * Two ways to specify the character pool, then draws `length` code points
 * from it uniformly:
 *
 *   - `<gen type="symbol" alphabet="kana.hiragana" length="8"/>` — a
 *     named alphabet from the registry.
 *   - `<gen type="symbol" value="कखगघ" length="1"/>` — an inline set,
 *     spelled directly (literals + `[x-y]` ranges, any Unicode). This is
 *     the friendly path that avoids regex for "pick one of these
 *     symbols".
 *
 * Exactly one of `alphabet` / `value` must be given. To pick a whole
 * WORD from a list, use `<gen type="text">` instead — `symbol` is
 * character-level.
 */

import { CharSetError, parseCharSet } from '../unicode/charset.js';
import { randomPick } from '../prng/random.js';
import { ALPHABET_NAMES, resolveAlphabetChars } from '../unicode/alphabets.js';

import type { Generator } from './generator.js';

export const DEFAULT_SYMBOL_LENGTH = 1;
export const MAX_SYMBOL_LENGTH = 1024;

export interface SymbolGenAttrs {
  readonly alphabet?: string | undefined;
  /** Inline character set, e.g. "कखगघ" or "[a-z][0-9]". */
  readonly value?: string | undefined;
  /**
   * Characters to ADD to the base set (same grammar as `value`).
   * Useful to extend a fixed named alphabet, e.g. alphabet="latin.lower"
   * include="[0-9]".
   */
  readonly include?: string | undefined;
  /**
   * Characters to REMOVE from the set (same grammar as `value`). Applied
   * last, so exclude has the final say — `value="[a-z]" exclude="y"`
   * yields a–z without y.
   */
  readonly exclude?: string | undefined;
  readonly length?: number | string | undefined;
}

export class SymbolGeneratorError extends Error {
  public override readonly name = 'SymbolGeneratorError';
}

export function parseSymbolLength(
  raw: number | string | undefined,
  fallback = DEFAULT_SYMBOL_LENGTH,
): number {
  if (raw === undefined) return fallback;
  const value = typeof raw === 'number' ? raw : Number(raw.trim());
  if (!Number.isSafeInteger(value) || value <= 0 || value > MAX_SYMBOL_LENGTH) {
    throw new SymbolGeneratorError(
      `symbol length must be an integer from 1 to ${String(MAX_SYMBOL_LENGTH)}, got "${String(raw)}"`,
    );
  }
  return value;
}

export function symbolGenerator(attrs: SymbolGenAttrs): Generator {
  const chars = resolveSymbolChars(attrs);
  const length = parseSymbolLength(attrs.length);
  return (count, prng) => {
    const out: string[] = new Array<string>(count);
    for (let i = 0; i < count; i++) {
      let value = '';
      for (let pos = 0; pos < length; pos++) value += randomPick(prng, chars);
      out[i] = value;
    }
    return out;
  };
}

/**
 * Resolve the character pool from either the inline `value` set or the
 * named `alphabet`. Exactly one must be provided.
 */
export function resolveSymbolChars(attrs: SymbolGenAttrs): readonly string[] {
  const value = attrs.value !== undefined && attrs.value.length > 0 ? attrs.value : undefined;
  const alphabet =
    attrs.alphabet !== undefined && attrs.alphabet.length > 0 ? attrs.alphabet : undefined;

  if (value !== undefined && alphabet !== undefined) {
    throw new SymbolGeneratorError(
      'symbol generator: use either "value" (inline set) or "alphabet" (named), not both',
    );
  }

  let base: readonly string[];
  if (value !== undefined) {
    const set = parseSetOrThrow(value, 'value');
    if (set.length === 0) {
      throw new SymbolGeneratorError(
        `symbol generator: value "${value}" produced an empty character set`,
      );
    }
    base = set;
  } else if (alphabet !== undefined) {
    const chars = resolveAlphabetChars(alphabet);
    if (chars === undefined) {
      throw new SymbolGeneratorError(
        `unknown alphabet "${alphabet}"; known alphabets: ${ALPHABET_NAMES.join(', ')}`,
      );
    }
    base = chars;
  } else {
    throw new SymbolGeneratorError(
      'symbol generator requires "value" (inline set like "abc" or "[a-z]") or "alphabet" (named); ' +
        `known alphabets: ${ALPHABET_NAMES.join(', ')}`,
    );
  }

  return applyIncludeExclude(base, attrs);
}

/**
 * Apply the `include` (add) and `exclude` (remove) modifiers to a base
 * set. Order: (base ∪ include) − exclude — exclude wins. Result preserves
 * order and stays de-duplicated.
 */
function applyIncludeExclude(base: readonly string[], attrs: SymbolGenAttrs): readonly string[] {
  const include = attrs.include;
  const exclude = attrs.exclude;
  if (
    (include === undefined || include.length === 0) &&
    (exclude === undefined || exclude.length === 0)
  ) {
    return base;
  }

  const seen = new Set(base);
  const chars: string[] = [...base];
  if (include !== undefined && include.length > 0) {
    for (const c of parseSetOrThrow(include, 'include')) {
      if (!seen.has(c)) {
        seen.add(c);
        chars.push(c);
      }
    }
  }

  let result = chars;
  if (exclude !== undefined && exclude.length > 0) {
    const removed = new Set(parseSetOrThrow(exclude, 'exclude'));
    result = chars.filter((c) => !removed.has(c));
  }

  if (result.length === 0) {
    throw new SymbolGeneratorError(
      'symbol generator: the character set is empty after applying include/exclude',
    );
  }
  return result;
}

function parseSetOrThrow(spec: string, label: string): string[] {
  try {
    return parseCharSet(spec);
  } catch (err) {
    if (err instanceof CharSetError) {
      throw new SymbolGeneratorError(`symbol ${label}: ${err.message}`);
    }
    throw err;
  }
}

File: rust/src/generators/symbol.rs
//! `<gen type="symbol" .../>` — characters, drawn one at a time.
//!
//! The pool comes from either a named `alphabet` or an inline `value` set, never
//! both. To pick a whole word from a list, `<gen type="text">` is the tag; this
//! one works below the level of words.

use std::collections::BTreeMap;
use std::collections::BTreeSet;

use super::rand;
use crate::engine::{invalid, EngineResult};
use crate::prng::Sfc32;
use crate::unicode::{self, char_set};

const DEFAULT_LENGTH: i32 = 1;
const MAX_LENGTH: i32 = 1024;

pub fn generate(
    attrs: &BTreeMap<String, String>,
    count: usize,
    prng: &mut Sfc32,
) -> EngineResult<Vec<String>> {
    let chars = resolve_chars(attrs)?;
    let length = parse_length(attrs.get("length").map(String::as_str))?;

    Ok((0..count)
        .map(|_| {
            (0..length)
                .map(|_| rand::pick(prng, &chars))
                .collect::<String>()
        })
        .collect())
}

pub fn parse_length(raw: Option<&str>) -> EngineResult<i32> {
    let Some(raw) = raw else {
        return Ok(DEFAULT_LENGTH);
    };
    match raw.trim().parse::<i32>() {
        Ok(v) if v > 0 && v <= MAX_LENGTH => Ok(v),
        _ => invalid(&format!(
            "symbol length must be an integer from 1 to {MAX_LENGTH}, got \"{raw}\""
        )),
    }
}

pub fn resolve_chars(attrs: &BTreeMap<String, String>) -> EngineResult<Vec<char>> {
    fn blank_to_none(s: Option<&String>) -> Option<&str> {
        s.map(String::as_str).filter(|v| !v.is_empty())
    }
    let value = blank_to_none(attrs.get("value"));
    let alphabet = blank_to_none(attrs.get("alphabet"));

    if value.is_some() && alphabet.is_some() {
        return invalid(
            "symbol generator: use either \"value\" (inline set) or \"alphabet\" (named), not both",
        );
    }

    let base = if let Some(value) = value {
        let parsed = char_set::parse(value)?;
        if parsed.is_empty() {
            return invalid(&format!(
                "symbol generator: value \"{value}\" produced an empty character set"
            ));
        }
        parsed
    } else if let Some(alphabet) = alphabet {
        match unicode::chars(alphabet) {
            Some(chars) => chars.to_vec(),
            None => {
                return invalid(&format!(
                    "unknown alphabet \"{alphabet}\"; known alphabets: {}",
                    unicode::names().join(", ")
                ))
            }
        }
    } else {
        return invalid(&format!(
            "symbol generator requires \"value\" (inline set like \"abc\" or \"[a-z]\") or \
             \"alphabet\" (named); known alphabets: {}",
            unicode::names().join(", ")
        ));
    };

    apply_include_exclude(
        base,
        attrs.get("include").map(String::as_str),
        attrs.get("exclude").map(String::as_str),
    )
}

/// `(base ∪ include) − exclude`. Exclude is applied last, so it always has the
/// last word.
fn apply_include_exclude(
    base: Vec<char>,
    include: Option<&str>,
    exclude: Option<&str>,
) -> EngineResult<Vec<char>> {
    let include = include.filter(|s| !s.is_empty());
    let exclude = exclude.filter(|s| !s.is_empty());
    if include.is_none() && exclude.is_none() {
        return Ok(base);
    }

    // First-seen order kept, because the set is indexed by a random draw.
    let mut order = base.clone();
    let mut seen: BTreeSet<char> = base.into_iter().collect();
    if let Some(include) = include {
        for c in char_set::parse(include)? {
            if seen.insert(c) {
                order.push(c);
            }
        }
    }
    if let Some(exclude) = exclude {
        let drop: BTreeSet<char> = char_set::parse(exclude)?.into_iter().collect();
        order.retain(|c| !drop.contains(c));
    }

    if order.is_empty() {
        return invalid(
            "symbol generator: the character set is empty after applying include/exclude",
        );
    }
    Ok(order)
}
